### DISCUSSION TITLE
set TURBO_FORCE when hitting a deploy webhook

### DISCUSSION THREAD

**Discussion by bobbyflowstate** [vercel/community#1170] [Help] [answered]
### Summary

Hi there! I'm trying to set the TURBO_FORCE flag when hitting a deploy webhook on my vercel project.

### Turbo Ignore Build Set
Right now, the default build setting has the following Ignore Build Step command: `npx turbo-ignore`. Hence when I hit the deploy webhook the build gets canceled because it detects no code changes. However the reason we're rebuilding (daily or weekly) is because the data backing our pages has changed. So while the code hasn't changed, we do in fact need to rebuild with new data.

### The TURBO_FORCE flag
When I manually select a build and choose **Redeploy** from the deployment page I see that a `TURBO_FORCE` flag is set:

```
≫   Using Turborepo to determine if this project is affected by the commit...
--
10:48:48.725 |  
10:48:48.725 | ≫   `TURBO_FORCE` detected
10:48:48.726 |  
10:48:48.726 | ✅  proceeding with deployment
10:48:49.885 | Running "vercel build"
```

I need to be able to set this flag when I hit the deploy webhook as well so we can programmatically trigger a (daily or weekly) forced build. Is this possible? Do you recommend another solution here? 

Appreciate your help!

**Comment by amyegan** [comment]
Hi @rajuashok! Can you share more about your project? Determining the best possible solutions will depend on the frameworks and libraries used in your app.

[Incremental Static Regeneration](https://vercel.com/docs/concepts/incremental-static-regeneration/overview#) is usually a good solution for pulling in content changes from a CMS without requiring a full site redeploy. If you are using Next.js or Web Output API, then the [ISR Quickstart](https://vercel.com/docs/concepts/incremental-static-regeneration/quickstart#) will help you get it set up.

**Reply by bobbyflowstate** [reply]
Thanks @amyegan for some of those ideas. We do plan to move towards ISR in the near future. However in the short term we need daily builds to accomplish new data appearing on our pages daily.

- Our stack is multiple Next.js projects within a single turborepo.
- We have several pages that are statically built and many more that are server side rendered.
- We also use pnpm as our package manager
- Let's say we want to build one of our apps daily within the following directory `/apps/sports`. This directory has a project within Vercel with the build command: `cd ../.. && npx turbo run build --filter=@app/sports`
- This also has an *Ignore Build Step* command with: `npx turbo-ignore` 
- When new code is committed to the entire repo that above command will check if the code affects the sports project and decide build or not. At least that's my understanding. I may be wrong 🫣
- We'd like to build daily even if the code has not changed hence the thought of using the `TURBO_FORCE` flag - or something similar.

Does this make sense? Let me know if you need more info.

**Reply by bobbyflowstate** [reply]
I just removed the *Ignore Build Step* command `npx turbo-ignore` and it's now building (even if there are no code changes) when I hit the deploy hook. This works for now, but isn't ideal.

**Reply by bobbyflowstate** [reply]
I have to admit the ISR setup is a lot easier than I initially thought. Might move to that sooner (after the holidays) than expected 🙃

**Reply by amyegan** [reply]
Thanks for clarifying! Because `turbo-ignore` has no way of knowing that a content change has been made, you're right that you would need to either force the build or remove that Ignore Build Step setting. There isn't a straightforward, reliable way to selectively force the build with deploy hooks or webhooks that I know of, so I think you have already found the simplest solution.

You could also run a manual [CLI deployment](https://vercel.com/docs/cli/deploying-from-cli) each day, or [deploy with GitHub Actions](https://docs.github.com/en/actions/deployment/about-deployments/deploying-with-github-actions#triggering-your-deployment) on a [schedule](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule) for a daily build. The guide on [how to use GitHub Actions with Vercel](https://vercel.com/guides/how-can-i-use-github-actions-with-vercel) walks through setting up a deployment action if you want to go with that option. But it involves lots of configuration that probably isn't worth the effort if you're planning to switch to ISR soon.

If I were you, I would keep what you're doing for now and move to ISR after the holidays, as you suggested. If you want to know more about the other options, just let me know and I'm happy to get you more information. Either way, I hope you have a wonderful holiday season!

**Reply by bobbyflowstate** [reply]
Thank you! Yes I'm gonna stick with my current solution and eventually switch to ISR. Really appreciate your help here - thank you and enjoy the holidays!

**Comment by bobbyflowstate** [comment, marked as answer]
FYI, I just found a solution to add the `TURBO_FORCE` flag to deploy webhooks! Simply append `?buildCache=false` to the deploy webhook and it will add the `TURBO_FORCE` flag hence causing the build to run even if `npx turbo-ignore` is set on the **IGNORE BUILD STEP** setting.

**Reply by bobbyflowstate** [reply]
For more infor: https://vercel.com/docs/concepts/git/deploy-hooks#build-cache

**Reply by amyegan** [reply]
TIL! Thanks for coming back to share this 😄
